lib/puppet/parser/compiler.rb

```diff
@@ -190,6 +207,44 @@ def evaluate_node_classes
     evaluate_classes(classes_without_params, @node_scope || topscope)
   end
 
+  def evaluate_applications
+    @applications.each do |app|
+      mapping = app.parameters[:nodes].value
+      components = []
+      mapping.each do |k,v|
+        k = [k] unless k.is_a?(Array)
+        v = [v] unless v.is_a?(Array)
+        k.each do |key|
+          v.each do |val|
+            key, val = val, key if val.type == "Node"
+            if key.type == "Node" && key.title == node.name
+              components << val
+            end
+          end
+        end
+      end
+      # @todo lutter 2015-01-28: We use @current_app and @curremt_comp to
+      # signal to the evaluator that we are in the middle of evaluating an
+      # application.
+      begin
+        @current_app = app
+        @current_components = components
```

Comment:
Initialized them to `nil` in `initialize` and added a comment explaining in more detail what they are used for.